File: java/WordSearch.java
import java.util.*;
public class Solution {
    public String word;
    public char[][] board;
    public boolean result = false;
    public boolean exist(char[][] board, String word) {
        this.board = board;
        this.word = word;
        int row = board.length;
        if (row == 0) {
            return this.word.length() == 0;
        }
        int col = board[0].length;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                this.dfs(i, j, 0);
            }
        }
        return this.result;
    }
    
    public void dfs(int row, int col, int curr) {
        if (this.result == true) {
            return;
        }
        if (this.word.charAt(curr) == this.board[row][col]) {
            if (curr == this.word.length()-1) {
                this.result = true;
                return;
            }
            char temp = this.board[row][col];
            this.board[row][col] = '.';
            if (row > 0) {
                this.dfs(row-1, col, curr+1);
            }
            if (row < this.board.length-1) {
                this.dfs(row+1, col, curr+1);
            }
            if (col > 0) {
                this.dfs(row, col-1, curr+1);
            }
            if (col < this.board[0].length-1) {
                this.dfs(row, col+1, curr+1);
            }
            this.board[row][col] = temp;
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        char[][] board = new char[1][1];
        board[0][0] = 'a';
        String s = "a";
        sol.exist(board, s);
        System.out.println(sol.result);
    }
}

File: archive/python/backtracking/WordSearch.py
class Solution:
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def __init__(self):
        self.board = []
        self.word = ""
        self.result = False

    def exist(self, board, word):
        self.word = word
        if len(self.word) == 0:
            return True

        simple_board = []
        for i in xrange(len(board)):
            simple_board.append(list(board[i]))
        self.board = simple_board

        for i in xrange(len(self.board)):
            for j in xrange(len(self.board[0])):
                if self.board[i][j] == self.word[0] and not self.result:
                    self.dfs(i, j, self.word[0])

        return self.result

    def dfs(self, row, col, s):
        temp = self.board[row][col]
        if len(s) == len(self.word):
            self.result = True
            return
        elif not self.result:
            self.board[row][col] = "."
            up = row-1
            down = row+1
            left = col-1
            right = col+1
            if up >= 0 and self.board[up][col] == self.word[len(s)]:
                self.dfs(up, col, s+self.board[up][col])
            if down < len(self.board) and self.board[down][col] == self.word[len(s)]:
                self.dfs(down, col, s+self.board[down][col])
            if left >= 0 and self.board[row][left] == self.word[len(s)]:
                self.dfs(row, left, s+self.board[row][left])
            if right < len(self.board[0]) and self.board[row][right] == self.word[len(s)]:
                self.dfs(row, right, s+self.board[row][right])
            self.board[row][col] = temp        

if __name__ == "__main__":
    solution = Solution()
    #print solution.exist(["aaa", "aaa", "aab"], "aaaaaaaab")

    #print solution.exist([["a"]], "a")
    #print solution.exist(["ab", "cd"], "acdb")
    #print solution.exist(["ab"], "ba")
    #print solution.exist(["CAA", "AAA", "BCD"], "AAB")
    import datetime
    print datetime.datetime.now()
    print solution.exist([
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "aaaaaaaaaaaaaaaaaaaaaaaaaaaaab"
        ],
        "baaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    )
    import datetime
    print datetime.datetime.now()
